src/ServiceControl.Audit/Auditing/AuditIngestionComponent.cs

```diff
@@ -64,9 +86,33 @@ public Task Start(IMessageSession session)
             return watchdog.Start();
         }
 
-        public Task Stop()
+        public async Task Stop()
         {
-            return watchdog.Stop();
+            await watchdog.Stop().ConfigureAwait(false);
+            channel.Writer.Complete();
+            await ingestionWorker.ConfigureAwait(false);
+        }
+
+        async Task Loop()
+        {
+            var contexts = new List<MessageContext>(settings.MaximumConcurrencyLevel);
+
+            while (await channel.Reader.WaitToReadAsync().ConfigureAwait(false))
+            {
+                try
+                {
+                    while (channel.Reader.TryRead(out var context))
```

Comment:
Comments here about what conditions lead to the TryRead returning false would be good. Is it bound by the in-flight messages semaphore? Does it return false if there's a lag in being able to read something (which would require another WaitToReadAsync to return true) and then what is that caused by? Are the batches here firmly constrained to a specific number?

Reply:
I've added a few comments. See if that is enough